kong/core/handler.lua

```diff
@@ -31,37 +37,197 @@ local function get_now()
 end
 
 
--- in the table below the `before` and `after` is to indicate when they run:
--- before or after the plugins
-return {
-  build_router = function()
-    local apis
+local function build_router(dao, version)
+  local apis, err = dao.apis:find_all()
+  if err then
+    return nil, "could not load APIs: " .. err
+  end
 
-    apis, router_err = singletons.dao.apis:find_all()
-    if router_err then
-      return nil, "could not load APIs: " .. router_err
+  for i = 1, #apis do
+    -- alias since the router expects 'headers' as a map
+    if apis[i].hosts then
+      apis[i].headers = { host = apis[i].hosts }
     end
+  end
 
-    for i = 1, #apis do
-      -- alias since the router expects 'headers'
-      -- as a map
-      if apis[i].hosts then
-        apis[i].headers = { host = apis[i].hosts }
-      end
-    end
+  router, err = Router.new(apis)
+  if not router then
+    return nil, "could not create router: " .. err
+  end
 
-    router, router_err = Router.new(apis)
-    if not router then
-      return nil, "could not create router: " .. router_err
-    end
+  if version then
+    router_version = version
+  end
+
+  return true
+end
 
-    return true
-  end,
+
+-- in the table below the `before` and `after` is to indicate when they run:
+-- before or after the plugins
+return {
+  build_router = build_router,
 
   init_worker = {
     before = function()
       reports.init_worker()
-      cluster.init_worker()
+
+      -- initialize local local_events hooks
+
+      local dao            = singletons.dao
+      local cache          = singletons.cache
+      local worker_events  = singletons.worker_events
+      local cluster_events = singletons.cluster_events
+
+
+      -- events dispatcher
+
+
+      worker_events.register(function(data)
+        if not data.schema then
+          log(ngx.ERR, "[events] missing schema in crud subscriber")
+          return
+        end
+
+        if not data.entity then
+          log(ngx.ERR, "[events] missing entity in crud subscriber")
+          return
+        end
+
+        -- invalidate this entity anywhere it is cached if it has a
+        -- caching key
+
+        local cache_key = dao[data.schema.table]:entity_cache_key(data.entity)
+        if cache_key then
+          cache:invalidate(cache_key)
+        end
+
+        -- if we had an update, but the cache key was part of what was updated,
+        -- we need to invalidate the previous entity as well
+
+        if data.old_entity then
+          cache_key = dao[data.schema.table]:entity_cache_key(data.old_entity)
+          if cache_key then
+            cache:invalidate(cache_key)
+          end
+        end
+
+        if not data.operation then
+          log(ngx.ERR, "[events] missing operation in crud subscriber")
+          return
+        end
+
+        local entity_channel           = fmt("%s", data.schema.table)
```

Comment:
Historically I had different channels, then updated them, but did not remove the `fmt` call. We can for sure :)